Improve documentation for Azure OpenAI, including Structured Outputs support

jdubois opened the pull request:
As discussed in https://github.com/langchain4j/langchain4j/pull/1982#issuecomment-2553095472

jdubois committed:
Improve documentation for Azure OpenAI, including Structured Outputs support

jdubois committed:
Merge branch 'main' into azure-openai-structured-outputs-doc

jdubois committed:
Merge branch 'main' into azure-openai-structured-outputs-doc

jdubois commented on docs/docs/integrations/language-models/azure-open-ai.md:
```diff
@@ -135,12 +139,128 @@ ChatLanguageModel model = AzureOpenAiChatModel.builder()
+    String functionCalling(Model model) {
+        String question = "Is the current Microsoft stock higher than $450?";
+
+        Assistant assistant = AiServices.builder(Assistant.class)
```
Oh yes, then I didn't want this to be a Spring Boot demo -> I'm going to rewrite this without Spring, to remove the confusion and have something generic

jdubois committed:
Merge branch 'main' into azure-openai-structured-outputs-doc

jdubois committed:
Remove Spring Boot in the Tools examples

jdubois committed:
Fix

jdubois committed:
Remove documentation for embeddings and images

jdubois commented on docs/docs/integrations/language-models/azure-open-ai.md:
````diff
@@ -171,19 +291,33 @@ langchain4j.azure-open-ai.streaming-chat-model.user-agent-suffix=...
 langchain4j.azure-open-ai.streaming-chat-model.customHeaders=...
 ```
 
+## Creating `AzureOpenAiEmbeddingModel` to manage embeddings
````
removed

jdubois commented on docs/docs/integrations/language-models/azure-open-ai.md:
```diff
@@ -171,19 +291,33 @@ langchain4j.azure-open-ai.streaming-chat-model.user-agent-suffix=...
 
-### Spring Boot
-The `AzureOpenAiTokenizer` bean is created automatically by the Spring Boot starter.
+## Creating `AzureOpenAiImageModel` to generate images
```
removes

jdubois committed:
Update docs/docs/tutorials/structured-outputs.md

Co-authored-by: LangChain4j <langchain4j@gmail.com>

jdubois committed:
Update docs/docs/tutorials/structured-outputs.md

Co-authored-by: LangChain4j <langchain4j@gmail.com>

jdubois committed:
Merge branch 'main' into azure-openai-structured-outputs-doc

jdubois committed:
add .supportedCapabilities(Set.of(RESPONSE_FORMAT_JSON_SCHEMA))

jdubois commented on docs/docs/tutorials/structured-outputs.md:
```diff
@@ -353,17 +353,25 @@ ChatLanguageModel chatModel = OpenAiChatModel.builder() // see [1] below
+        .endpoint(System.getenv("AZURE_OPENAI_URL"))
+        .apiKey(System.getenv("AZURE_OPENAI_API_KEY"))
+        .modelName("gpt-4o-mini")
+        .strictJsonSchema(true) // see [3] below
```
Done

jdubois committed:
Add documentation for AzureOpenAiChatModel in the structured outputs

jdubois commented on docs/docs/tutorials/structured-outputs.md:
```diff
@@ -353,17 +353,25 @@ ChatLanguageModel chatModel = OpenAiChatModel.builder() // see [1] below
         .logResponses(true)
         .build();
 // OR
+ChatLanguageModel chatModel = AzureOpenAiChatModel.builder() // see [1] below
```
Done

jdubois commented on docs/docs/integrations/language-models/azure-open-ai.md:
```diff
@@ -135,12 +139,128 @@ ChatLanguageModel model = AzureOpenAiChatModel.builder()
+If `strictJsonSchema` is set to `false` and you provide a JSON Schema, the model will still try to generate a response that adheres to the schema, but it will not fail if the response does not adhere to the schema. One reason to do this is for better performance.
+:::
+
+Once the model is configured, the JSON Schema has to be specified in the `ChatRequest` object for each request.
```
My issue is that I'm trying to be neutral with the framework (=not depend on Spring too heavily, in case people don't use it, and also as its auto configuration magic makes it a bit harder to explain what's going on).
@Tool comes from LangChain4j "core", so there's no issue with it, but @AiService comes with the Spring Boot starter.

dliubarskyi replied:
@jdubois I see. There is no need to use `@AiService`, you can create `Assistant` instance manually, but still profit from the `Person` as a return type. Automatically generating json schema is not Spring-specific, only `@AiService` is

jdubois replied:
OK, I've totally rewritten this part, and I now have documentation for both:
- the high level API is indeed simpler and more consistent with the documentation above on Tools
- the low level API allows to understand better how the JSON schema works

jdubois commented:
@langchain4j can you have a look at the only review comment that's left? Otherwise this should be good to go!

dliubarskyi commented:
@jdubois great, thanks!

BTW I have made @langchain4j an organization, so please use @dliubarskyi to ping me

jdubois committed:
Detail both the high level and the low level API for structured outputs

jdubois commented:
Great @dliubarskyi ! I've just fixed the last comment and updated my own demo that showcases it at https://github.com/jdubois/jdubois-langchain4j-demo
Now there's a LangChain4j org, can I join it?

dliubarskyi commented on docs/docs/integrations/language-models/azure-open-ai.md:
```diff
@@ -78,6 +80,8 @@ langchain4j.azure-open-ai.chat-model.stop=...
 langchain4j.azure-open-ai.chat-model.presence-penalty=...
 langchain4j.azure-open-ai.chat-model.frequency-penalty=...
 langchain4j.azure-open-ai.chat-model.seed=...
+langchain4j.azure-open-ai.chat-model.response-format=...
```
```suggestion
```
There is no such propery

dliubarskyi commented on docs/docs/tutorials/structured-outputs.md:
```diff
@@ -353,17 +360,26 @@ ChatLanguageModel chatModel = OpenAiChatModel.builder() // see [1] below
+        .apiKey(System.getenv("AZURE_OPENAI_API_KEY"))
+        .deploymentName("gpt-4o-mini")
+        .strictJsonSchema(true) // see [3] below
+        .supportedCapabilities(Set.of(RESPONSE_FORMAT_JSON_SCHEMA))
```
```suggestion
        .supportedCapabilities(Set.of(RESPONSE_FORMAT_JSON_SCHEMA)) // see [3] below
```

dliubarskyi commented on docs/docs/tutorials/structured-outputs.md:
```diff
@@ -353,17 +360,26 @@ ChatLanguageModel chatModel = OpenAiChatModel.builder() // see [1] below
+        .endpoint(System.getenv("AZURE_OPENAI_URL"))
+        .apiKey(System.getenv("AZURE_OPENAI_API_KEY"))
+        .deploymentName("gpt-4o-mini")
+        .strictJsonSchema(true) // see [3] below
```
```suggestion
        .strictJsonSchema(true)
```

dliubarskyi commented on docs/docs/integrations/language-models/azure-open-ai.md:
```diff
@@ -135,23 +139,182 @@ ChatLanguageModel model = AzureOpenAiChatModel.builder()
+        .endpoint(System.getenv("AZURE_OPENAI_URL"))
+        .apiKey(System.getenv("AZURE_OPENAI_KEY"))
+        .deploymentName("gpt-4o")
+        .strictJsonSchema(true)
```
```suggestion
        .strictJsonSchema(true)
        .supportedCapabilities(Set.of(RESPONSE_FORMAT_JSON_SCHEMA))
```

dliubarskyi commented on docs/docs/integrations/language-models/azure-open-ai.md:
````diff
@@ -135,23 +139,182 @@ ChatLanguageModel model = AzureOpenAiChatModel.builder()
+```java
+
+interface PersonAssistant {
+    Person favoriteColor(String message);
````
```suggestion
    Person extractPerson(String message);
```

dliubarskyi commented on docs/docs/integrations/language-models/azure-open-ai.md:
```diff
@@ -135,23 +139,182 @@ ChatLanguageModel model = AzureOpenAiChatModel.builder()
+                .chatLanguageModel(chatLanguageModel)
+                .build();
+
+Person person = assistant.favoriteColor(question);
```
```suggestion
Person person = assistant.extractPerson(question);
```

dliubarskyi commented on docs/docs/integrations/language-models/azure-open-ai.md:
```diff
@@ -135,23 +139,182 @@ ChatLanguageModel model = AzureOpenAiChatModel.builder()
+    Person favoriteColor(String message);
+}
+
+class Person {
```
nit: could have been a `record`

dliubarskyi commented on docs/docs/integrations/language-models/azure-open-ai.md:
````diff
@@ -135,23 +139,182 @@ ChatLanguageModel model = AzureOpenAiChatModel.builder()
+
+```java
+ChatRequest chatRequest = ChatRequest.builder()
+    .messages(singletonList(userMessage("Julien likes the colors blue, white and red")))
````
```suggestion
    .messages(UserMessage.from("Julien likes the colors blue, white and red"))
```

dliubarskyi reviewed (dismissed):
@jdubois thank you!

dliubarskyi committed:
Update docs/docs/integrations/language-models/azure-open-ai.md

dliubarskyi committed:
Update docs/docs/tutorials/structured-outputs.md

dliubarskyi committed:
Update docs/docs/tutorials/structured-outputs.md

dliubarskyi committed:
Update docs/docs/integrations/language-models/azure-open-ai.md

dliubarskyi committed:
Update docs/docs/integrations/language-models/azure-open-ai.md

dliubarskyi committed:
Update docs/docs/integrations/language-models/azure-open-ai.md

dliubarskyi committed:
Update docs/docs/integrations/language-models/azure-open-ai.md

dliubarskyi commented on docs/docs/integrations/language-models/azure-open-ai.md:
```diff
@@ -135,23 +138,183 @@ ChatLanguageModel model = AzureOpenAiChatModel.builder()
+:::
+
+You can then use this model either with the high level `Assistant` API or the low level `ChatLanguageModel` API, as detailed below.
+
```
```suggestion
When using it with the high level `Assistant` API, configure `supportedCapabilities(Set.of(RESPONSE_FORMAT_JSON_SCHEMA))` to enable structured outputs with a JSON schema.

```

dliubarskyi committed:
Update docs/docs/integrations/language-models/azure-open-ai.md

dliubarskyi commented:
@jdubois 

> Now there's a LangChain4j org, can I join it?

What kind of permissions do you require?

jdubois commented:
Thanks for merging @dliubarskyi !!! For permissions, I'd love to be able to manage issues: this would be only for assigning myself or other people on the issues that are tagged with Azure, of course.
